src/w3iProxy/externalCommunicators/jsCommunicator.ts

```diff
@@ -21,7 +21,23 @@ export class JsCommunicator implements ExternalCommunicator {
       const messageListener = (messageResponse: JsonRpcResult<TReturn>) => {
         resolve(messageResponse.result)
       }
+
       this.emitter.once(message.id.toString(), messageListener)
+
+      const externalMessageListener = (messageEvent: MessageEvent) => {
+        console.log('GOT MESSAGE LAW IT', messageEvent.data)
+        if (messageEvent.data.id === message.id || messageEvent.data.id === message.id.toString()) {
+          if (messageEvent.data.result) {
+            window.removeEventListener('message', externalMessageListener)
+            resolve(messageEvent.data.result)
+          }
+        }
+      }
+
+      window.addEventListener('message', externalMessageListener)
+
+      // Emit all three to allow for debugging and widget usage
```

Comment:
+1 for more contextual comments 🎉 

Why three? I'm counting two effective emits (postMessage, emit)